### ISSUE TITLE
Can't compile using AOCC Compilers

### ISSUE THREAD

**Issue by ericnardi** [HDFGroup/hdf5#280]
I'm trying to compile the latest stable release (1.12.0) on a Linux machine, using [AMD Optimizing C/C++ Compilers](https://developer.amd.com/amd-aocc/), but I get errors when running make.

The configure step completes succesfully (I have zlib compiled with AOCC and installed to the directory pointed by the PREFIX variable):

```
./configure --prefix=$PREFIX --with-zlib=$PREFIX --enable-fortran
```

But when running make, I get the following error:

```
  FCLD     libhdf5_fortran.la
clang-11: error: unknown argument: '-soname'
clang-11: error: no such file or directory: 'libhdf5_fortran.so.200'
make[3]: *** [Makefile:963: libhdf5_fortran.la] Error 1
make[3]: Leaving directory '/shared/home/enardi/src/hdf5-1.12.0/fortran/src'
make[2]: *** [Makefile:877: all] Error 2
make[2]: Leaving directory '/shared/home/enardi/src/hdf5-1.12.0/fortran/src'
make[1]: *** [Makefile:828: all-recursive] Error 1
make[1]: Leaving directory '/shared/home/enardi/src/hdf5-1.12.0/fortran'
make: *** [Makefile:662: all-recursive] Error 1
```

**Comment by byrnHDF** [comment]
From that error log it indicates that the fortran library did not get built and the setting of the so name failed because it could not find the file. Therefore there must be another failure in either the compile of the sources or the linking of the library.

**Comment by ericnardi** [comment]
Hi @byrnHDF, thanks for replying.

Yes, actually, it seems like the fortran shared library did not get built because the -soname flag is not identified by this version of the clang compiler, and the way I see it these flags should actually be sent to the linker in this case (via -Wl flags). It seems that libtool doesn't recognize these AOCC compilers very well and these flags are incorrectly set. I have searched around a bit and the developers of FLang already [discussed](https://github.com/flang-compiler/clang/issues/15#issue-235288021) about a similar issue.

For now, I think I'll try manually running the commands to build the shared library and see if Make can continue from there.

**Comment by byrnHDF** [comment]
Any possibility to try building with CMake, to see if that system can handle this compiler?

**Comment by ericnardi** [comment]
Good idea, I just tried it, but no luck...

But now the problem seems different, maybe I am missing something. I get the following errors when trying to compile with CMake:

```
[ 26%] Linking C executable ../bin/use_append_chunk
cd /shared/home/enardi/src/hdf5-build/test && /shared/opt/gnusoft/bin/cmake -E cmake_link_script CMakeFiles/use_append_chunk.dir/link.txt --verbose=1
/shared/opt/aocc-compiler-2.3.0/bin/clang  -std=c99 -fPIC -O3 -DNDEBUG -L/shared/opt/amdsoft/lib CMakeFiles/use_append_chunk.dir/use_append_chunk.c.o CMakeFiles/use_append_chunk.dir/use_common.c.o -o ../bin/use_append_chunk  -Wl,-rpath,"\$ORIGIN/../lib:\$ORIGIN/" ../bin/libhdf5_test.so.200.0.0 -lm ../bin/libhdf5.so.200.0.0 -ldl
ld.lld: error: ../bin/libhdf5.so.200.0.0: undefined reference to compress2 [--no-allow-shlib-undefined]
ld.lld: error: ../bin/libhdf5.so.200.0.0: undefined reference to inflate [--no-allow-shlib-undefined]
ld.lld: error: ../bin/libhdf5.so.200.0.0: undefined reference to inflateEnd [--no-allow-shlib-undefined]
ld.lld: error: ../bin/libhdf5.so.200.0.0: undefined reference to inflateInit_ [--no-allow-shlib-undefined]
clang-11: error: linker command failed with exit code 1 (use -v to see invocation)
gmake[2]: *** [test/CMakeFiles/use_append_chunk.dir/build.make:120: bin/use_append_chunk] Error 1
gmake[2]: Leaving directory '/shared/home/enardi/src/hdf5-build'
gmake[1]: *** [CMakeFiles/Makefile2:2692: test/CMakeFiles/use_append_chunk.dir/all] Error 2
gmake[1]: Leaving directory '/shared/home/enardi/src/hdf5-build'
gmake: *** [Makefile:182: all] Error 2
```

And I tried to manually run the build commands, but 'make' doesn't see the compiled library, it just tries to recompile it and fails with the same error.

**Comment by BtbN** [comment]
I just ran into this myself, and the issue really is libtool/configure horribly misdetecting flang.

To manually fix it, open the generated libtool script (right next to the configure script) after you ran configure.

First, search for `CC="clang++"` and take note of the values it found there.
Then search for `CC="flang"` and see that it found nothing/nonsense for the same values. Write in the same values it found for clang++.

In my case, it looked like this after I was done editing:

```
# A language specific compiler.
CC="flang"

# Is the compiler the GNU compiler?
with_gcc=yes

# Compiler flag to turn off builtin functions.
no_builtin_flag=" -fno-builtin"

# Additional compiler flags for building library objects.
pic_flag=" -fPIC -DPIC"

# How to pass a linker flag through the compiler.
wl="-Wl,"

# Compiler flag to prevent dynamic linking.
link_static_flag="-static"

# Does compiler simultaneously support -c and -o options?
compiler_c_o="yes"

# Whether or not to add -lc for building shared libraries.
build_libtool_need_lc=no

# Whether or not to disallow shared libs when runtime libs are static.
allow_libtool_libs_with_static_runtimes=no
```

The values below this don't seem to matter/are correct.

After that, the build with AOCC completed successfully.


While at it, another issue I ran into with AOCC was that configure failed, trying to add "-lwp" to the linker flags.
That's because it sees `-target-feature -lwp` in the linker invocation from flang -v, and decides that it must be a library.
To work around it, manually set FCLIBS.
Right before configure throws the error, it prints out the contents of that variable. Just copy it, but remove -lwp from it, and invoke configure again:

`FCLIBS="..." ./configure ...`

**Comment by derobins** [comment]
AOCC works with develop, @lrknox will check to make sure this is true for all maintenance branches

**Comment by lcebaman** [comment]
@derobins Are you sure? I am facing problems with develop:
` ld.lld: error: can't create dynamic relocation R_X86_64_32S against local symbol in readonly segment; recompile object files with -fPIC or pass '-Wl,-z,notext' to allow text relocations in the output`

**Comment by hyoklee** [comment]
#3657

**Comment by derobins** [comment]
This is/was a libtool error. The libtool script that is generated at compile time was missing the "-Wl," prefix necessary for linker options, so "-Wl,-soname" gets parsed as -soname, which fails.

As of October 2023, with AOCC 4.1, c/flang works out of the box w/ shared unless you are using an MPI compiler wrapper. We have a hack for that (at least in the Autotools), which will appear shortly.

This issue will be closed when the PR is merged to develop.

**Comment by derobins** [comment]
Fixed w/ #3674
